Language InterOp Blog based on cppyy and Cling paper from 2023 (#167)
compiler-research/compiler-research.github.io: _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md

---

> title: "The Next Step in Language Interoperability using Cling and cppyy"

Reviewer: ```suggestion
title: "The Next Step in Language Interoperability using compiler-as-a-service and cppyy"
```

layout: post
excerpt: "Scientific software in constantly being challenged by enthusiasts trying to
test the boundaries of programming languages, in search of better performance

> and simpler workflows. One such breakthrough was achieved using Cling, a C++
> interpreter, that has presented new possibilities with an incremental
> compilation infrastructure that is available at runtime."

Reviewer: ```suggestion
and simpler workflows. Interactive C++ interpreters such as Cling and ClangRepl
presented new possibilities with an incremental compilation infrastructure that
 is available at runtime."
```

sitemap: false
permalink: blogs/language_interoperability_with_cling_and_cppyy/
date: 2023-04-05
---

{% capture image_style %}
    max-width: 80%;
    display: block;
    margin: 0 auto;
{% endcapture %}

Scientific software in constantly being challenged by enthusiasts trying to
test the boundaries of programming languages, in search of better performance
and simpler workflows. One such breakthrough was achieved using Cling, a C++
interpreter, that has presented new possibilities with an incremental

> compilation infrastructure that is available at runtime.

Reviewer: Replace the this as suggested above.

This means that Python can interact with C++ on an on-demand basis, and
bindings can be automatically constructed at runtime. This provides
unprecedented performance and does not require direct support from library
authors.

The Compiler Research team presented these findings in the paper: [Efficient
and Accurate Automatic Python Bindings with cppyy & Cling]. It presents the
enhancements in language interoperability using Cling with cppyy (an
automatic, run-time, Python-C++ bindings generator). Following is a high-level
summary of these findings.

### Background
C++ gained early adoption in scientific research fields due to its high
performance capabilities. But the interactive nature of Python and the gentler
learning curve led to higher adoption rates elsewhere, and as a result, it saw
exponential advancements in capabilities that were ideal for data science

> research. However, this does not discredit the usefulness of C++ in data
> science. A lot of research infrastructure is still rooted in C++ and benefits
> from some of its unique features (e.g., access to accelerators in
> heterogeneous computing environments). 

Reviewer: ```suggestion
research. Python shines when steering infrastructure written in high
performance language such as C or C++. However, it is challenging to
write the glue layers between both languages for every package available
in the C++ ecosystem.
```

This is where the usefulness of language interoperability becomes evident.
However, this requires an advanced integration solution, especially for high
performance code that can suffer from penalties crossing the language barrier.

![numba extension](/images/blog/cppyy-numba-1.png){: style="{{ image_style }}"}

Numba, a just-in-time (JIT) compiler for Python, is a tool that is ideal for
this task (with some enhancements). Numba is capable of compiling Python code
while targeting either the CPU or the GPU, and providing interfaces to use the
JITed closures from low-level libraries. Numba helps lower Python to machine
code level and minimizes costly language crossings. In order to provide the

> missing links for this research, Numba was also integrated with cppyy (an
> automatic runtime bindings generator).

Reviewer: ```suggestion
missing links for this research, Numba was also integrated with the cppyy project -- an
automatic runtime bindings generator based using interactive C++ to connect to the Python runtime.
```
